app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt

```diff
@@ -0,0 +1,168 @@
+package org.oppia.app.parser
+
+import org.oppia.app.model.CurrencyUnit
+import org.oppia.app.model.Fraction
+import org.oppia.app.model.NumberUnit
+import org.oppia.app.model.NumberWithUnits
+import org.oppia.app.model.Units
+import org.oppia.domain.util.normalizeWhitespace
+import java.lang.Double
+
+import java.util.regex.Pattern
+
+/** This class contains methods that help to parse string to number, check realtime and submit time errors. */
+class StringToNumberWithUnitsParser {
+  private lateinit var value: String
+  private lateinit var units: String
+  private lateinit var CURRENCY_UNITS: HashMap<String, Units>
+
+  /** @return index of pattern in s or -1, if not found
+   */
+  fun indexOf(pattern: Pattern, s: String): Int {
+    val matcher = pattern.matcher(s)
+    return if (matcher.find()) matcher.start() else -1
+  }
+
+  fun getCurrencyUnits(): CurrencyUnit {
```

Comment:
Let's initialize this inline above, and we should be using ``listOf()`` and ``mapOf()`` to keep immutability. See the following initial example:

```kotlin
private const val CURRENCY_UNITS = mapOf(
  "dollar" to Unit(name = "dollar", aliases = listOf("$", "dollars", ...), frontUnits = listOf("$")),
  ...
)
```

Note that I intend for these to be defined outside the class since it's actually a constant.

Reply:
Changed to
 `companion object {
     private val CURRENCY_UNITS = mapOf(
      "dollar" to Units(name = "dollar",...),...)
}`